spring-session/src/main/java/org/springframework/session/SafeRetrievingSessionRepository.java

```diff
@@ -0,0 +1,162 @@
+ * scenarios where {@link SessionRepository} consumer wants to treat the session that
+ * wasn't deserializable as non-existing.
+ * <p>
+ * By default, this implementation will delete the session whose retrieval using
+ * {@link #getSession(String)} has failed with the configured ignored exception. This
+ * behavior can be changed using {@link #setDeleteOnIgnoredException(boolean)} method.
+ *
+ * @param <S> the type of {@link Session} being managed
+ * @author Vedran Pavic
+ * @since 1.3.0
+ */
+public class SafeRetrievingSessionRepository<S extends Session>
+		implements FindByIndexNameSessionRepository<S> {
+
+	private final FindByIndexNameSessionRepository<S> delegate;
+
+	private final Set<Class<? extends RuntimeException>> ignoredExceptions;
+
+	private boolean deleteOnIgnoredException = true;
+
+	/**
+	 * Create a new {@link SafeRetrievingSessionRepository} instance backed by a
+	 * {@link FindByIndexNameSessionRepository} delegate.
+	 * @param delegate the {@link FindByIndexNameSessionRepository} delegate
+	 * @param ignoredExceptions the set of exceptions to ignore
+	 */
+	public SafeRetrievingSessionRepository(
+			FindByIndexNameSessionRepository<S> delegate,
+			Set<Class<? extends RuntimeException>> ignoredExceptions) {
+		Assert.notNull(delegate, "Delegate must not be null");
+		Assert.notEmpty(ignoredExceptions, "Ignored exceptions must not be empty");
+		this.delegate = delegate;
+		this.ignoredExceptions = ignoredExceptions;
+	}
+
+	/**
+	 * Create a new {@link SafeRetrievingSessionRepository} instance backed by a
+	 * {@link SessionRepository} delegate.
+	 * @param delegate the {@link SessionRepository} delegate
+	 * @param ignoredExceptions the set of exceptions to ignore
+	 */
+	public SafeRetrievingSessionRepository(SessionRepository<S> delegate,
+			Set<Class<? extends RuntimeException>> ignoredExceptions) {
+		this(new FindByIndexNameSessionRepositoryAdapter<S>(delegate), ignoredExceptions);
+	}
+
+	/**
+	 * Set whether session should be deleted after ignored exception has occurred during
+	 * retrieval.
+	 * @param deleteOnIgnoredException the flag to indicate whether session should be
+	 * deleted
+	 */
+	public void setDeleteOnIgnoredException(boolean deleteOnIgnoredException) {
+		this.deleteOnIgnoredException = deleteOnIgnoredException;
+	}
+
+	public S createSession() {
+		return this.delegate.createSession();
+	}
+
+	public void save(S session) {
+		this.delegate.save(session);
+	}
+
+	public S getSession(String id) {
+		try {
+			return this.delegate.getSession(id);
+		}
+		catch (RuntimeException e) {
+			if (isIgnoredException(e)) {
+				if (this.deleteOnIgnoredException) {
+					this.delegate.delete(id);
+				}
+				return null;
+			}
+			throw e;
+		}
+	}
+
+	public void delete(String id) {
+		this.delegate.delete(id);
+	}
+
+	public Map<String, S> findByIndexNameAndIndexValue(String indexName, String indexValue) {
+		try {
+			return this.delegate.findByIndexNameAndIndexValue(indexName, indexValue);
```

Comment:
Thank you for raising this concern @shakuzen - I've planned to comment on `findByIndexNameAndIndexValue` behavior but forgot. I'm not sure what would be the best way to address this due to exact reasons you've described. Delegates will fail even if only one of the results is not deserializable and I don't see a way around it.